src/Mod/PartDesign/App/FeaturePad.cpp

```diff
@@ -115,18 +129,62 @@ App::DocumentObjectExecReturn *Pad::execute(void)
         base = TopoDS_Shape();
     }
 
-
     // get the Sketch plane
-    Base::Placement SketchPos    = obj->Placement.getValue();
-    Base::Vector3d  SketchVector = getProfileNormal();
+    Base::Placement SketchPos = obj->Placement.getValue();
+    // get the normal vector of the sketch
+    Base::Vector3d SketchVector = getProfileNormal();
 
     try {
+        // get user-defined decimals
+        int precision = Base::UnitsApi::getDecimals();
+
         this->positionByPrevious();
         TopLoc_Location invObjLoc = this->getLocation().Inverted();
 
         base.Move(invObjLoc);
 
-        gp_Dir dir(SketchVector.x,SketchVector.y,SketchVector.z);
+        Base::Vector3d paddingDirection;
+
+        // use the given vector if necessary
+        if (!UseCustomVector.getValue()) {
+            paddingDirection.x = SketchVector.x;
+            paddingDirection.y = SketchVector.y;
+            paddingDirection.z = SketchVector.z;
+        }
+        else {
+            // if null vector, use SketchVector
+            if ( (RoundTo(XSkew.getValue(), precision) == 0.0)
+                && (RoundTo(YSkew.getValue(), precision) == 0.0)
+                && (RoundTo(ZSkew.getValue(), precision) == 0.0) ) {
+                XSkew.setValue(SketchVector.x);
+                YSkew.setValue(SketchVector.y);
+                ZSkew.setValue(SketchVector.z);
+            }
+            paddingDirection.x = XSkew.getValue();
+            paddingDirection.y = YSkew.getValue();
+            paddingDirection.z = ZSkew.getValue();
+        }
+
+        // create vector in padding direction with length 1
+        gp_Dir dir(paddingDirection.x, paddingDirection.y, paddingDirection.z);
+
+        // The length of a gp_Dir is 1 so the resulting pad would have
+        // the length L in the direction of dir. But we want to have its height in the
+        // direction of the normal vector.
+        // Therefore we must multiply L by the factor that is necessary
+        // to make dir as long that its projection to the SketchVector
+        // equals the SketchVector.
+        // This is the scalar product of both vectors.
+
+        double factor = dir * gp_Dir(SketchVector.x, SketchVector.y, SketchVector.z);
+
+        // factor would be zero if vectors are orthogonal
+        if (RoundTo(factor, precision) == 0.0)

```

Comment:
again, don't rely on user settings, so that recomputing is not affected by FreeCAD UI settings. Use Precision::Angular() instead. Otherwise, different people may have different results on their different freecads when opening and recomputing one same file.